src/memory/redis.rs

```diff
@@ -0,0 +1,253 @@
+            .iter()
+            .flat_map(|&value| value.to_le_bytes().to_vec())
+            .collect();
+
+        let result: redis::Value = redis::cmd("FT.SEARCH")
+            .arg(&self.index_name) // Replace with the actual index name
+            .arg("DIALECT")
+            .arg(2)
+            .arg(format!("($query)=>[KNN {} @v $B]", min_count))
+            .arg("PARAMS")
+            .arg("query")
+            .arg(&query_blob)
+            .query_async(&mut con)
+            .await?;
+
+        let result_pairs: Vec<(String, f32)> = match result {
+            redis::Value::Bulk(items) => {
+                items
+                    .chunks_exact(2)
+                    .filter_map(|chunk| match (chunk.get(0), chunk.get(1)) {
+                        (Some(redis::Value::Data(key)), Some(redis::Value::Data(value))) => {
+                            let score: f32 = String::from_utf8_lossy(value)
+                                .parse()
+                                .unwrap_or_default();
+                            Some((String::from_utf8_lossy(key).into_owned(), score))
+                        }
+                        _ => None,
+                    })
+                    .collect()
+            }
+            _ => vec![],
+        };
+
+        let mut relevant_memories = vec![];
+        for (key, similarity) in result_pairs {
+            let json_data: String = redis::cmd("JSON.GET")
+                .arg(&key)
+                .query_async(&mut con)
+                .await?;
+            let data: serde_json::Value = serde_json::from_value(serde_json::Value::String(json_data))?;
+            let memory_data: RedisPayload = serde_json::from_value(data["memory"].clone())?;
+            let memory_embedding: Vec<f32> = serde_json::from_value(data["embedding"].clone())?;
+
+            relevant_memories.push(RelevantMemory {
+                memory: Memory {
+                    content: memory_data.content,
+                    recall: memory_data.recall,
+                    recency: memory_data.recency,
+                    embedding: memory_embedding,
+                },
+                relevance: similarity,
+            });
+        }
+
+        // Sort the relevant memories by relevance and return the top min_count memories
+        relevant_memories.sort_by(|a, b| b.relevance.partial_cmp(&a.relevance).unwrap_or(Ordering::Equal));
+        Ok(relevant_memories.into_iter().take(min_count).collect())
+    }
+}
+
+pub struct RedisProvider;
+
+impl MemoryProvider for RedisProvider {
+    fn is_enabled(&self) -> bool {
+        true
+    }
+
+    fn get_name(&self) -> String {
+        "redis".to_string()
+    }
+
+    fn create(&self, _: Value) -> Result<Box<dyn MemorySystem> ,Box<dyn Error> > {
+        let client = Client::open("redis://127.0.0.1/")?;
+
+        let rt = Runtime::new().expect("Failed to create Tokio runtime");
+
+        let index_name = "smartgpt_agent_memory";
```

Comment:
Similar to Qdrasnt memory, this should be a configrable option. I think you made an option for that.